### PR TITLE
New payment flow: one-off strip payment

### PR DESCRIPTION
**Pull request by regiskuckaertz**
## Why are you doing this?

Adds the logic to wire money for single contributions via Stripe. There was some logic I could reuse from the existing code, with a few modifications. Rather than addressing these changes locally and risk a regression, I just copy-pasted the code elsewhere—I don't think it belongs there anyway. The aim is that it becomes the only logic necessary to handle payments via Stripe (single, monthly and annual included); then later a lot of code can be deleted. Ok, so what's going on here:

- I am focusing on one-off contribution via Stripe first and will generalise in a follow-up PR
- The server-side logic is updated to reflect that: at the moment, it requires the same logic as the one-off contribution code (_with_ the user data if available)
- I have updated the Redux state type so that it includes user and csrf data, although I don't have a clue yet how the latter needs to be used
- I moved the form into a bona fide component (`ContributionForm`) so that it can receive properties like any other
- I copied the content of [ajax.js](https://github.com/guardian/support-frontend/blob/master/assets/pages/oneoff-contributions/helpers/ajax.js) into [stripeCheckout.js](https://github.com/guardian/support-frontend/blob/master/assets/pages/oneoff-contributions/helpers/ajax.js) with a few modifications that make it a little bit more decoupled—they are documented in situ.

I expect some more refactoring when I tackle recurring contributions with Stripe. In particular, all the pre-/post- logic should be moved into its own set of helpers, it should be independent of any payment method.

TODO
- [ ] use csrf properly
- [ ] add routing logic for switching between the form and the thank you page
- [ ] write `onSuccess` logic
- [ ] write `onError` logic

[**Trello Card**](https://trello.com/c/U5Uu58QI/715-npf-stripe-one-off)

### CODE REVIEW

**Comment by regiskuckaertz** [comment]
It seems `csrf` isn't used in the current one-off payment logic, so I'll remove for now.

File: assets/helpers/paymentIntegrations/stripeCheckout.js [hunk 1]
```diff
@@ -1,13 +1,149 @@
...
+  return Promise.resolve({
+    ok: false,
+    text: () => 'Failed to process payment - missing fields',
+  });
+}
+
+function postToEndpoint(request: Object, onSuccess: Function, onError: Function): Promise<*> {
+  return fetch(stripeOneOffContributionEndpoint(cookie.get('_test_username')), request).then((response) => {
+    if (response.ok) {
+      onSuccess();
```
Comment (regiskuckaertz) [inline assets/helpers/paymentIntegrations/stripeCheckout.js]: `onSuccess` should only be called after all error conditions have been tested.

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+          <NewContributionAmount
+            countryGroupId={countryGroupId}
+            countryGroupDetails={selectedCountryGroupDetails}
+            currency={currency}
+          />
+          <NewContributionTextInput 
+            id="contributionFirstName" 
+            name="contribution-fname" 
+            label="First Name" 
+            value={window.guardian.user && window.guardian.user.firstName ? window.guardian.user.firstName : null}
```
Comment (regiskuckaertz) [inline assets/pages/new-contributions-landing/components/ContributionForm.jsx]: I was told this was the existing behaviour, although I did not find anything such in the one-off logic.

File: assets/helpers/paymentIntegrations/stripeCheckout.js [hunk 3]
```diff
@@ -1,13 +1,147 @@
...
 
 // ----- Functions ----- //
 
+function requestData(
+  abParticipations: Participations,
+  paymentToken: string,
+  currency: IsoCurrency,
+  referrerAcquisitionData: ReferrerAcquisitionData,
+  optimizeExperiments: OptimizeExperiments,
+  getData: Function,
```
Comment (Mullefa) [inline assets/helpers/paymentIntegrations/stripeCheckout.js]: What's the reason that you're not using more specific types for `getData`, `request`, `onError` and `onSuccess`?

File: assets/helpers/tracking/conversions.js [hunk 4]
```diff
@@ -9,10 +9,12 @@ export default function trackConversion(
   participations: Participations,
   currentRoute: string,
 ) {
-  // Fire GTM conversion events
-  successfulConversion(participations);
-  // Send an Ophan pageview. Because this function is used to track page views
-  // from client side routed thank you pages, the referrer will always be the current location
-  pageView(getAbsoluteURL(currentRoute), document.location.href);
+  if (navigator.doNotTrack !== "1") {
```
Comment (Mullefa) [inline assets/helpers/tracking/conversions.js]: There's a `doNotTrack` function which returns a boolean

File: assets/helpers/paymentIntegrations/stripeCheckout.js [hunk 5]
```diff
@@ -1,13 +1,147 @@
...
+) {
+  const { user, amount } = getData();
```
Comment (Mullefa) [inline assets/helpers/paymentIntegrations/stripeCheckout.js]: What is the advantage of this, as opposed to fixing the amounts in the callback?

File: assets/helpers/paymentIntegrations/stripeCheckout.js (hunk 3 above)
Comment (regiskuckaertz) [reply assets/helpers/paymentIntegrations/stripeCheckout.js]: Great question! I initially had some problems but I think I can now define a proper type 👍

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx (hunk 2 above)
Comment (jranks123) [reply assets/pages/new-contributions-landing/components/ContributionForm.jsx]: It is in here: https://github.com/guardian/support-frontend/blob/master/assets/pages/oneoff-contributions/components/formFields.jsx

```<TextInput
        id="name"
        placeholder="Full name"
        labelText="Full name"
        value={props.fullName.value}
        onChange={props.setFullName}
        onBlur={onFormFieldBlur(props.setFullNameShouldValidate, stripeButtonClassName)}
        modifierClasses={['name']}
        showError={shouldShowError(props.fullName)}
        errorMessage="Please enter your name."
      />```

Basically the props.fullName is initialised in the user state as `""`, and then if there is a user in the window, the user reducer overwrites `""` with the users name

File: assets/helpers/paymentIntegrations/stripeCheckout.js (hunk 5 above)
Comment (regiskuckaertz) [reply assets/helpers/paymentIntegrations/stripeCheckout.js]: The callback is setup at rendering time, while it is called on submit. Since I am not using Redux state to store the form data, I need to read it off the form elements, which must happen on submit.

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx (hunk 2 above)
Comment (jranks123) [reply assets/pages/new-contributions-landing/components/ContributionForm.jsx]: Infact, looking at the file, I think rather than reading from the window, we should be passing in these values as props to the component

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+  abParticipations: Participations,
+  referrerAcquisitionData: ReferrerAcquisitionData,
+  optimizeExperiments: OptimizeExperiments,
+  thankYouRoute: string,
+  onSuccess: () => void,
+  onError: string => void,
+|};
+/* eslint-enable react/no-unused-prop-types */
+
+const mapStateToProps = (state: State) => ({
```
Comment (jranks123) [inline assets/pages/new-contributions-landing/components/ContributionForm.jsx]: You can read the values for first name, last name, email etc from the user state here. 

So something like:

```
const mapStateToProps = (state: State) => ({
  done: state.page.form.done,
  isTestUser: state.page.user.isTestUser || false,
  referrerAcquisitionData: state.common.referrerAcquisitionData,
  abParticipations: state.common.abParticipations,
  optimizeExperiments: state.common.optimizeExperiments,
  firstName: state.page.user.firstName,
  lastName: state.page.user.lastName,
  email: state.page.user.email,
});
```

That way you don't need to read from the `window.guardian` object further down

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+          <NewContributionAmount
+            countryGroupId={countryGroupId}
+            countryGroupDetails={selectedCountryGroupDetails}
+            currency={currency}
+          />
+          <NewContributionTextInput
+            id="contributionFirstName"
+            name="contribution-fname"
+            label="First Name"
+            value={window.guardian.user && window.guardian.user.firstName ? window.guardian.user.firstName : null}
```
Comment (jranks123) [inline assets/pages/new-contributions-landing/components/ContributionForm.jsx]: won't this make the input's uneditable? as the value of `window.guardian.user.firstName` will never change

File: assets/helpers/paymentIntegrations/stripeCheckout.js (hunk 5 above)
Comment (Mullefa) [reply assets/helpers/paymentIntegrations/stripeCheckout.js]: I see. Thanks for explaining.

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx (hunk 7 above)
Comment (jranks123) [reply assets/pages/new-contributions-landing/components/ContributionForm.jsx]: I've just tried this locally and it isn't the case, and it's because there isn't anything on the form at the moment that changes the state yet, which would cause a re-render and write over anything written in the boxes. This is why we have to use the state to set the value of inputs

Comment (regiskuckaertz) [reply assets/pages/new-contributions-landing/components/ContributionForm.jsx]: Nope. Thanks to the DOM diffing, these elements will never be re-rendered. However I agree that in certain circumstances you'd be forced to keep track of the field values 😭

File: app/controllers/Application.scala [hunk 8]
```diff
@@ -97,26 +102,26 @@ class Application(
...
-    Ok(views.html.main(
-      title = "Support the Guardian | Thank you for your contribution",
-      description = None,
-      mainId = s"new-contributions-thank-you-page-$countryCode",
-      mainJsBundle = "newThankYouPage.js",
-      mainStyleBundle = "newContributionsLandingPageStyles.css"
-    ))
-  }
+  private def newContributions(countryCode: String, idUser: Option[IdUser])(implicit request: RequestHeader) =
+    views.html.oneOffContributions(
```
Comment (joelochlann) [inline app/controllers/Application.scala]: I think it would be great to create a new dedicated "skeleton" page right now, since it's confusing seeing `oneOffContributions` being used as the shell for a page which can do both one-off and recurring

File: assets/helpers/tracking/conversions.js [hunk 9]
```diff
@@ -9,8 +9,10 @@ export default function trackConversion(
   participations: Participations,
   currentRoute: string,
 ) {
-  // Fire GTM conversion events
-  successfulConversion(participations);
+  if (navigator.doNotTrack !== '1') {
```
Comment (joelochlann) [inline assets/helpers/tracking/conversions.js]: Not really related to your code but it's worrying to me that (as you mentioned in person) you had to add this code in because an error in `successfulConversion` was stopping execution of further code

@Ap0c @rupertbates this makes me think we're missing `try/catch` around the contents of this function? unless I'm missing something?

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx (hunk 7 above)
Comment (jranks123) [reply assets/pages/new-contributions-landing/components/ContributionForm.jsx]: What are these circumstances? If we don't need to then that's great! I'll have to get you to explain  this DOM-diffing to me tomorrow

File: assets/helpers/paymentIntegrations/stripeCheckout.js [hunk 10]
```diff
@@ -1,13 +1,155 @@
...
+
+  return {
+    method: 'POST',
+    headers: { 'Content-Type': 'application/json' },
+    body: JSON.stringify(oneOffContribFields),
+    credentials: 'include',
+  };
+}
+
+function postToEndpoint(request: Object, onSuccess: Function, onError: Function): Promise<*> {
```
Comment (joelochlann) [inline assets/helpers/paymentIntegrations/stripeCheckout.js]: Could we make use of the fact that this returns a promise rather than passing in callbacks? We could `throw` on the error conditions and the caller could supply a `.catch()`

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx (hunk 7 above)
Comment (regiskuckaertz) [reply assets/pages/new-contributions-landing/components/ContributionForm.jsx]: In short, yes. I just think it is a terrible design on the part of React/Redux to impose that kind of constraint, but it doesn't mean I will bend to it 😉 There is no reason to duplicate a perfectly working state machine (the form) on the client side, that's why these properties are immutable (there is alas no such thing in JS except through discipline). Duplication leads to complecting (having to deal with `onchange` and `onblur`) and bugs like the forever hanging page that happened the other day.

File: app/controllers/Application.scala (hunk 8 above)
Comment (regiskuckaertz) [reply app/controllers/Application.scala]: Yes I'm doing that.

**Comment by joelochlann** [comment]
It might be worth waiting til tomorrow to merge though, in case of post-deployment test failures

File: assets/helpers/paymentIntegrations/stripeCheckout.js (hunk 10 above)
Comment (regiskuckaertz) [reply assets/helpers/paymentIntegrations/stripeCheckout.js]: Hmm, I don't think so. This logic is happening _after_ stripe checkout, it is invoked via the supplied `token` callback in `Stripe.configure` ([here](https://github.com/guardian/support-frontend/blob/8f41da5a95dad01da0218456a8abb0b7eca714cd/assets/helpers/paymentIntegrations/stripeCheckout.js#L213))

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx (hunk 7 above)
Comment (regiskuckaertz) [reply assets/pages/new-contributions-landing/components/ContributionForm.jsx]: > What are these circumstances?

When the rendering of the element is conditional, you can either abstain from supplying a value or you are forced to track the value in the state. So in this case:
```
{condition ? <input value={intialValue} /> : null }
```
That's fair enough, as there's an easy workaround:
```
<input value={initialValue} hidden={!condition} />
```

**Comment by regiskuckaertz** [comment]
I'm sick of that Codacy thing, that error message does not mean anything.

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    thankYouRoute,
+  } = props;
+
+  const callback = createTokenCallback({
+    abParticipations,
+    currencyId: currency,
+    referrerAcquisitionData,
+    optimizeExperiments,
+    getData: () => {
+      const elements: any = formElement.elements; // eslint-disable-line prefer-destructuring
```
[the remark this answers is not in the document]
Comment (regiskuckaertz) [reply assets/pages/new-contributions-landing/components/ContributionForm.jsx]: The problem is `HTMLFormControlsCollection` isn't recognized by Flow and so I'm forced into casting it to `any` ... and I don't know how to marry that with destructuring 😅

File: assets/pages/new-contributions-landing/components/ContributionForm.jsx [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+  const amount = getAmount(elements);
+  const email = elements.namedItem('contributionEmail').value;
+
+  openDialogBox(amount, email);
+};
+
+
+// ----- Render ----- //
+
+function setupStripe(formElement, props) {
```
Comment (jranks123) [inline assets/pages/new-contributions-landing/components/ContributionForm.jsx]: could we add types here?
